Fix mango full text detection

davisp opened the pull request:
The check for whether full text support in mango is available was broken
in that the mango_cursor_text module is only compiled if dreyfus exists
while the check to use the module was a runtime check. Thus is a user
were to add dreyfus after mango had been compiled it would have led to a
runtime error when we attempted to use a module that didn't exist.

This changes the check to a compile time check in both places. And now
that its a compile time define we can remove the need to rename source
files around which causes problems with source control seeing that file
change depending on local configuration.

COUCHDB-3378

nickva commented:
Get this error

```
ERROR: Failed to load /Users/nvatama/asf2/src/mango/rebar.config: {error,
                                                                   {26,file,
                                                                    {error,
                                                                     undef,
                                                                     [{lists,
                                                                       keystore,
                                                                       [erl_opts,
                                                                        1,
                                                                        {erl_opts,
                                                                         [{d,
                                                                           'HAVE_DREYFUS'}]}],
                                                                       []},
                                                                      {erl_eval,
                                                                       do_apply,
                                                                       6,
                                                                       [{file,
                                                                         "erl_eval.erl"},
                                                                        {line,
                                                                         670}]},
                                                                      {file,
                                                                       eval_stream2,
                                                                       6,
                                                                       [{file,
                                                                         "file.erl"},
                                                                        {line,
                                                                         1386}]},
                                                                      {file,
                                                                       script,
                                                                       2,
                                                                       [{file,
                                                                         "file.erl"},
                                                                        {line,
                                                                         1086}]},
                                                                      {rebar_config,
                                                                       new,2,
                                                                       [{file,
                                                                         "src/rebar_config.erl"},
                                                                        {line,
                                                                         238}]},
                                                                      {rebar_core,
                                                                       process_dir,
                                                                       4,
                                                                       [{file,
                                                                         "src/rebar_core.erl"},
                                                                        {line,
                                                                         131}]},
                                                                      {rebar_core,
                                                                       process_each,
                                                                       5,
                                                                       [{file,
                                                                         "src/rebar_core.erl"},
                                                                        {line,
                                                                         351}]},
                                                                      {rebar_core,
                                                                       process_dir1,
                                                                       7,
                                                                       [{file,
                                                                         "src/rebar_core.erl"},
                                                                        {line,
                                                                         253}]}]}}}

```

This seems to fix the issue:

```
 lists:keystore(erl_opts, 1, CONFIG, {erl_opts, NewOpts}),
```

davisp commented:
Yeap. Silly mistake.

Also went back and double checked how our build system does Dreyfus work and realized that the relative file path is quite dumb since people may want to be able to pull dreyfus in without having ot modify couchdb in a build chain. I managed to find a much cleaner check for dreyfus so updating that as well.

davisp commented:
Updated. Fixed the silly lists:keystore mistake and replaced the relative paths with a `code:lib_dir(dreyfus)` check instead which is more robust than requiring dreyfus in a specific place on the filesystem.

nickva commented:
Default:

```
1> l(mango_cursor).
{module,mango_cursor}
2> mango_cursor:module_info().
[{module,mango_cursor},
 {exports,[{create,3},
           {explain,1},
           {execute,3},
           {maybe_filter_indexes,2},
           {module_info,0},
           {module_info,1}]},
 {attributes,[{vsn,[144027518958084404948992969280201332310]}]},
 {compile,[{options,[{outdir,"ebin"},
                     debug_info,debug_info,
                     {i,"../"},
                     {i,"include"}]},
           {version,"7.0.3"},
           {source,"/Users/nvatama/asf2/src/mango/src/mango_cursor.erl"}]},
 {native,false},
 {md5,<<108,90,176,239,182,81,16,121,233,213,38,200,90,
        22,150,86>>}]
```

with Dreyfus dep:

```
diff --git a/rebar.config.script b/rebar.config.script
index 201cc0c0e..a1e516dc8 100644
--- a/rebar.config.script
+++ b/rebar.config.script
@@ -57,6 +57,8 @@ DepDescs = [
 %% Non-Erlang deps
 {docs,             {url, "https://github.com/apache/couchdb-documentation"},
                    "3b15a7cb5e1958bc71ac865b56d3081ab43f5942", [raw]},
+{dreyfus,          {url, "git@github.com:cloudant-labs/dreyfus"},
+                  "960725070dc75916d8ed4a24e8a7bbc208cbe1cd"},
 {fauxton,          {url, "https://github.com/apache/couchdb-fauxton"},
                   {tag, "v1.1.11"}, [raw]},
 %% Third party deps
```

```
1> l(mango_cursor).
{module,mango_cursor}
2> mango_cursor:module_info().
[{module,mango_cursor},
 {exports,[{create,3},
           {explain,1},
           {execute,3},
           {maybe_filter_indexes,2},
           {module_info,0},
           {module_info,1}]},
 {attributes,[{vsn,[65408884502798159025533969682876367424]}]},
 {compile,[{options,[{outdir,"ebin"},
                     debug_info,
                     {d,'HAVE_DREYFUS'},
                     {i,"include"}]},
           {version,"7.0.3"},
           {source,"/Users/nvatama/asf2/src/mango/src/mango_cursor.erl"}]},
 {native,false},
 {md5,<<49,53,74,253,105,250,209,249,34,62,28,25,96,208,
        130,64>>}]
```

nickva commented:
+1
